transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/connector/FragmentationIT.java

```diff
@@ -0,0 +1,1029 @@
+
+        context.checking(new Expectations() {
+            {
+                oneOf(handler).sessionCreated(with(any(IoSessionEx.class)));
+                oneOf(handler).sessionOpened(with(any(IoSessionEx.class)));
+                allowing(handler).messageReceived(with(any(IoSessionEx.class)), with(any(Object.class)));
+                will(new CustomAction("Capture messageReceived() parameters") {
+                    @Override
+                    public Object invoke(Invocation invocation) throws Throwable {
+                        Object message = invocation.getParameter(1);
+                        IoBufferEx ioBuffer = (IoBufferEx) message;
+                        bufferList.add(ioBuffer.buf());
+
+                        WsnSession wsnConnectSession = (WsnSession) invocation.getParameter(0);
+                        if (wsnConnectSession != null) {
+                            IoFilterChain filterChain = wsnConnectSession.getFilterChain();
+                            IoBufferAllocatorEx<? extends WsBuffer> allocator = wsnConnectSession.getBufferAllocator();
+
+                            final boolean hasPostUpgradeChildWsnSession = wsnConnectSession.getHandler() == handler;
+                            final ResourceAddress wsnSessionLocalAddress = wsnConnectSession.getLocalAddress();
+                            final boolean isLightweightWsnSession = wsnSessionLocalAddress.getOption(LIGHTWEIGHT);
+                            boolean sendMessagesDirect = isLightweightWsnSession
+                                                         && hasPostUpgradeChildWsnSession; // post-upgrade
+
+                            if (sendMessagesDirect) {
+                                filterChain.fireMessageReceived(message);
+                                return null;
+                            }
+
+                            if (bufferList.size() == 5) {
+                                ByteBuffer buffer = ByteBuffer.allocate(4096);
+
+                                for (ByteBuffer bb : bufferList) {
+                                    buffer.put(bb);
+                                }
+
+                                buffer.flip();
+                                WsBuffer wsBuffer = allocator.wrap(buffer, IoBufferEx.FLAG_SHARED);
+                                wsBuffer.setKind(WsBuffer.Kind.BINARY);
+                                wsnConnectSession.write(wsBuffer);
+                            }
+                        }
+
+                        return null;
+                    }
+                });
+                oneOf(handler).exceptionCaught(with(any(IoSessionEx.class)), with(any(Throwable.class)));
+                will(new CustomAction("Capture exception") {
+                    @Override
+                    public Object invoke(Invocation invocation) throws Throwable {
+                        Throwable throwable = (Throwable) invocation.getParameter(1);
```

Comment:
Make sure the test will fail if exceptionCaught is called. You could use a count down latch to signal all is finished from the handler (when it's got all the expected messages) and maybe an atomic reference or array to set an exception so you can assert it's null at the end of the test.